### PR TITLE
MvcControllersWithContext

### PR DESCRIPTION
**Pull request by prafullbhosale**

### CODE REVIEW

File: src/SampleWebApp/Models/AuthorDataContext.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using SampleWebApp.Models;
+
+namespace SampleWebApp.Data
+{
+    public class AuthorDataContext : DbContext
```
Comment (divega) [inline src/SampleWebApp/Models/AuthorDataContext.cs:10]: Is it an explicit decision to call these generated classes "Data Context"? Not sure it is a good idea to come up with yet another variation on what to call one of these on an already crowded space.

File: src/SampleWebApp/Models/BookDataContext.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using SampleWebApp.Models;
+
+namespace SampleWebApp.Data
+{
+    public class BookDataContext : DbContext
```
Comment (divega) [inline src/SampleWebApp/Models/BookDataContext.cs:10]: Is it expected that scaffolding would generate two separate DbContexts for the two classes?

File: src/SampleWebApp/Models/AuthorDataContext.cs (hunk 1 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Models/AuthorDataContext.cs:10]: The class name is user provided. 
The suggested name that is shown to the users while creating a new Context is <ApplicationName>Context

File: src/SampleWebApp/Models/BookDataContext.cs (hunk 2 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Models/BookDataContext.cs:10]: No, this is also decided by the user. 
Scaffolding allows to generate a controller for a single model per invocation. 

In this case I ran scaffolding twice, and for each of those I chose to create a new DbContext. 

If the user chooses an existing DbContext, scaffolding will only attempt to add the DbSet<> property in the existing DbContext class.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        public AuthorsMvcController(AuthorDataContext context)
+        {
+            _context = context;    
+        }
+
+        // GET: AuthorsMvc
+        public IActionResult Index()
+        {
+            return View(_context.Author.ToList());
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:26]: Not async? All of these should be async, no?

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+        // GET: AuthorsMvc/Details/5
+        public IActionResult Details(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            Author author = _context.Author.SingleOrDefault(m => m.AuthorID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:37]: Use `var`

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+        // GET: AuthorsMvc/Edit/5
+        public IActionResult Edit(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            Author author = _context.Author.SingleOrDefault(m => m.AuthorID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:74]: `var`

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs (hunk 3 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:26]: The async is a flag that needs to be passed in. 
The UI for scaffolding has a checkbox. If its checked, all these would be async.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        // POST: AuthorsMvc/Edit/5
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public IActionResult Edit(Author author)
+        {
+            if (ModelState.IsValid)
+            {
+                _context.Update(author);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:89]: I'm wondering whether we should we handle concurrency exceptions by default here and return an appropriate result.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        // GET: AuthorsMvc/Delete/5
+        [ActionName("Delete")]
+        public IActionResult Delete(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            Author author = _context.Author.SingleOrDefault(m => m.AuthorID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:105]: `var`

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+            return View(author);
+        }
+
+        // POST: AuthorsMvc/Delete/5
+        [HttpPost, ActionName("Delete")]
+        [ValidateAntiForgeryToken]
+        public IActionResult DeleteConfirmed(int id)
+        {
+            Author author = _context.Author.SingleOrDefault(m => m.AuthorID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:119]: `var`

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        // POST: AuthorsMvc/Create
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public IActionResult Create(Author author)
+        {
+            if (ModelState.IsValid)
+            {
+                _context.Author.Add(author);
```
Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:59]: Just call Add on the context here. Or be consistent and call Update on the set in the code below. The behavior is the same either way.

File: src/SampleWebApp/Startup.cs [hunk 10]
```diff
@@ -41,6 +42,11 @@ public Startup(IHostingEnvironment env, IApplicationEnvironment appEnv)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddDbContext<AuthorDataContext>(options =>
+                    options.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=AuthorDataContext-9a1d8034-a6d6-4cdb-bf59-77e85aad38ac;Trusted_Connection=True;MultipleActiveResultSets=true"));
+
+            services.AddDbContext<BookDataContext>(options =>
+                    options.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=BookDataContext-586e0d4b-10df-4ce2-b3a7-8d98621d6dcf;Trusted_Connection=True;MultipleActiveResultSets=true"));
```
Comment (ajcvickers) [inline src/SampleWebApp/Startup.cs:49]: Are we okay with scaffolding connection strings, which may have sensitive data, directly into the code?

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs (hunk 3 above)
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:26]: I had commented on the API one first, but I'll repeat what I said there... I'm not sure there's any reason to default to sync - when we were doing this before, async was a _new_ feature and plenty of people didn't grok it. Now it seems like async should be the default.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.EntityFrameworkCore;
+
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:8]: This one has a blank line here between framework and user namespaces, API controller did not.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        // GET: AuthorsMvc/Create
+        public IActionResult Create()
+        {
+            return View();
+        }
+
+        // POST: AuthorsMvc/Create
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public IActionResult Create(Author author)
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:55]: We used to have code to generate `[Bind(...)]` with all the known properties to prevent overposting.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            {
+                return NotFound();
+            }
+            return View(author);
+        }
+
+        // POST: AuthorsMvc/Edit/5
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public IActionResult Edit(Author author)
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:85]: This doesn't check that the `author.AuthorId` is equal to the route value and the API version does.

File: src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+<h2>Delete</h2>
+
+<h3>Are you sure you want to delete this?</h3>
+<div>
+    <h4>Author</h4>
+    <hr />
+    <dl class="dl-horizontal">
+        <dt>
+            @Html.DisplayNameFor(model => model.FirstMidName)
```
Comment (rynowak) [inline src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml:15]: @DamianEdwards - concerned about this tag-helping enough?

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+        public BooksMvcController(BookDataContext context)
+        {
+            _context = context;    
+        }
+
+        // GET: BooksMvc
+        public IActionResult Index()
+        {
+            var bookDataContext = _context.Book.Include(b => b.Author);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:26]: This seems like a really strange variable name. Shouldn't it be something like `var books`?

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+        // GET: BooksMvc/Details/5
+        public IActionResult Details(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            Book book = _context.Book.SingleOrDefault(m => m.BookID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:38]: `var`

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+                return NotFound();
+            }
+
+            return View(book);
+        }
+
+        // GET: BooksMvc/Create
+        public IActionResult Create()
+        {
+            ViewData["AuthorID"] = new SelectList(_context.Set<Author>(), "AuthorID", "Author");
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:50]: Why are we using `ViewData`? I know the select helpers have magic logic in them to use it but I personally **hate** it. Can we please make it scaffold explicit stuff instead?

Also, use `nameof(Author.AuthorID)`, etc. instead of strings for property names.

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+        // GET: BooksMvc/Edit/5
+        public IActionResult Edit(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            Book book = _context.Book.SingleOrDefault(m => m.BookID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:77]: `var`

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+                _context.Update(book);
+                _context.SaveChanges();
+                return RedirectToAction("Index");
+            }
+            ViewData["AuthorID"] = new SelectList(_context.Set<Author>(), "AuthorID", "Author", book.AuthorID);
+            return View(book);
+        }
+
+        // GET: BooksMvc/Delete/5
+        [ActionName("Delete")]
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:102]: Why is this needed for this method?

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+        public IActionResult Delete(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            Book book = _context.Book.SingleOrDefault(m => m.BookID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:110]: `var`

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+            return View(book);
+        }
+
+        // POST: BooksMvc/Delete/5
+        [HttpPost, ActionName("Delete")]
+        [ValidateAntiForgeryToken]
+        public IActionResult DeleteConfirmed(int id)
+        {
+            Book book = _context.Book.SingleOrDefault(m => m.BookID == id);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:124]: `var`

File: src/SampleWebApp/Controllers/BooksMvcController.cs [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+            _context.Book.Remove(book);
```
Comment (DamianEdwards) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:125]: Do we really need to do the full SELECT -> TRACK -> DELETE cycle here? Can we do something more like the update case above so you don't have to incur a select to perform the delete?

File: src/SampleWebApp/Models/AuthorDataContext.cs [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public DbSet<Author> Author { get; set; }
```
Comment (DamianEdwards) [inline src/SampleWebApp/Models/AuthorDataContext.cs:12]: properties after constructors

File: src/SampleWebApp/Startup.cs (hunk 10 above)
Comment (DamianEdwards) [inline src/SampleWebApp/Startup.cs:49]: Doesn't seem in line with what our templates by default.

File: src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml (hunk 14 above)
Comment (DamianEdwards) [inline src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml:15]: GAH! Tag Helpers please.

File: src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    ViewData["Title"] = "Delete";
+}
+
+<h2>Delete</h2>
+
+<h3>Are you sure you want to delete this?</h3>
+<div>
+    <h4>Author</h4>
+    <hr />
+    <dl class="dl-horizontal">
```
Comment (DamianEdwards) [inline src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml]: We don't use `<dl>` for form layout in the templates anymore. Please make it match https://github.com/aspnet/Templates/blob/dev/src/Rules/StarterWeb/IndividualAuth/Views/Account/Login.cshtml

File: src/SampleWebApp/Views/AuthorsMvc/Index.cshtml [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+    ViewData["Title"] = "Index";
+}
+
+<h2>Index</h2>
+
+<p>
+    <a asp-action="Create">Create New</a>
+</p>
+<table class="table">
+    <tr>
```
Comment (DamianEdwards) [inline src/SampleWebApp/Views/AuthorsMvc/Index.cshtml:13]: `<table>` should have `<thead>`, `<tbody>`, etc.

File: src/SampleWebApp/Views/BooksMvc/Create.cshtml [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+            </div>
+        </div>
+    </div>
+</form>
+
+<div>
+    <a asp-action="Index">Back to List</a>
+</div>
+
+@section Scripts {
```
Comment (DamianEdwards) [inline src/SampleWebApp/Views/BooksMvc/Create.cshtml:60]: The project template has a partial for the validation scripts, and I believe the old scaffolding had a checkbox that allowed the user to say whether they wanted this block generated or not, as it's very common for people to just put this block in the layout page instead. We need to figure out what we want to do this time around.

Also, we should be rendering debug scripts in Development, and min in Production, etc., like we do in the partial (https://github.com/aspnet/Templates/blob/dev/src/Rules/StarterWeb/IndividualAuth/Views/Shared/_ValidationScriptsPartial.cshtml).

File: src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml (hunk 14 above)
Comment (DamianEdwards) [inline src/SampleWebApp/Views/AuthorsMvc/Delete.cshtml:15]: Actually this is just display name so HTML helpers are fine. Carry on :smile:

File: src/SampleWebApp/Views/BooksMvc/Edit.cshtml [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+
+<h2>Edit</h2>
+
+<form asp-action="Edit">
+    <div class="form-horizontal">
+        <h4>Book</h4>
+        <hr />
+        <div asp-validation-summary="ValidationSummary.ModelOnly" class="text-danger"></div>
+        <div class="form-group">
+            <label asp-for="AuthorID" class="control-label col-md-2">AuthorID</label>
```
Comment (DamianEdwards) [inline src/SampleWebApp/Views/BooksMvc/Edit.cshtml:15]: The element shouldn't have a the value here (i.e. `AuthorID` shouldn't be between the tags).

File: src/SampleWebApp/Views/BooksMvc/Details.cshtml [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        </dd>
+        <dt>
+            @Html.DisplayNameFor(model => model.Year)
+        </dt>
+        <dd>
+            @Html.DisplayFor(model => model.Year)
+        </dd>
+    </dl>
+</div>
+<p>
```
Comment (DamianEdwards) [inline src/SampleWebApp/Views/BooksMvc/Details.cshtml:39]: This block of links is a `<p>` in some cases and a `<div>` in others. Please make it consistent.

File: src/SampleWebApp/Views/BooksMvc/Index.cshtml [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+    ViewData["Title"] = "Index";
+}
+
+<h2>Index</h2>
+
+<p>
+    <a asp-action="Create">Create New</a>
+</p>
+<table class="table">
+    <tr>
```
Comment (DamianEdwards) [inline src/SampleWebApp/Views/BooksMvc/Index.cshtml:13]: Same `<table>` comments from above.

File: src/SampleWebApp/Views/BooksMvc/Create.cshtml (hunk 26 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Views/BooksMvc/Create.cshtml:60]: This option is still available to the user. This block gets generated only if the 'Reference Script libraries' is checked. 

image

File: src/SampleWebApp/Views/BooksMvc/Edit.cshtml (hunk 27 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Views/BooksMvc/Edit.cshtml:15]: This was a bug in the scaffolding template's tag helper. Its addressed as part of aspnet/Scaffolding/issues/189

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs (hunk 6 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:89]: Will add the catch block for DbUpdateConcurrencyException similar to what we have in the ApiControllerWithContext

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs (hunk 9 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:59]: will change to `_context.Add(author);`

File: src/SampleWebApp/Startup.cs (hunk 10 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Startup.cs:49]: This is a known issue from RC1. We do plan to move the connection strings to appsettings.json for RTW.

File: src/SampleWebApp/Controllers/AuthorsMvcController.cs (hunk 11 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorsMvcController.cs:8]: Will remove the blank line to make them consistent.

File: src/SampleWebApp/Controllers/BooksMvcController.cs (hunk 15 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:26]: The current name is based on the DbContext name. 
I agree that the current naming is weird, but getting to `books` from the type name `Book` is plurarlization which we currently do not have.

File: src/SampleWebApp/Views/BooksMvc/Edit.cshtml (hunk 27 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Views/BooksMvc/Edit.cshtml:15]: My bad, this wasn't fixed earlier. 
The reason this was here was for foreign keys to be able to display the association name(AuthorID -> Author), but is currently unimplemented, so it always shows up as foreign key itself. 

Will remove it till we have a proper way to determine the associationName.

File: src/SampleWebApp/Controllers/BooksMvcController.cs (hunk 15 above)
Comment (rowanmiller) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:26]: I would use the DbSet property name, which in this example is `Book`, but in most real apps would be `Books` (BTW that is what we do for naming tables starting in RC2).

Comment (rynowak) [inline src/SampleWebApp/Controllers/BooksMvcController.cs:26]: @prafullbhosale - does the dialog let you customize the DbSet property name when generating the code? This might be something we want to consider. IIRC we didn't support this in the past, but it might make sense to have for EF Core
